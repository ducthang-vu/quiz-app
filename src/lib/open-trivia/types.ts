/**
 * Code 0: Success Returned results successfully.
 * Code 1: No Results Could not return results. The API doesn't have enough quiz for your query. (Ex. Asking for 50 Questions in a Category that only has 20.)
 * Code 2: Invalid Parameter Contains an invalid parameter. Arguments passed in aren't valid. (Ex. Amount = Five)
 * Code 3: Token Not Found Session Token does not exist.
 * Code 4: Token Empty Session Token has returned all possible quiz for the specified query. Resetting the Token is necessary.
 * Code 5: Rate Limit Too many requests have occurred. Each IP can only access the API once every 5 seconds.
 */
export enum ResponseCode {
    SUCCESS = 0,
    NO_RESULTS = 1,
    INVALID_PARAMETER = 2,
    TOKEN_NOT_FOUND = 3,
    TOKEN_EMPTY = 4,
    RATE_LIMIT = 5
}

export interface BaseResponse {
    response_code: ResponseCode;
}

export type GameDifficulty = 'any_difficulty' | 'easy' | 'medium' | 'hard';

export type GameType = 'multiple' | 'boolean' | 'both';

export interface TokenResponse extends BaseResponse {
    response_message: string;
    token: string;
}

export interface Question {
    category: string;
    type: GameType;
    difficulty: GameDifficulty;
    question: string;
    correct_answer: string;
    incorrect_answers: string[];
}

export interface QuestionsResponse extends BaseResponse {
    results: Question[];
}

export interface GetQuestionsParams {
    amount: number;
    category?: string;
    difficulty: GameDifficulty;
    type: GameType;
    token: string;
}
