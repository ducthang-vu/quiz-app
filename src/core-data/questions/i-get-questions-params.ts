import { NumberQuestions } from './number-questions';
import { GameDifficulty } from '../game-difficulty';
import { GameType } from '../game-type';

export interface IGetQuestionsParams {
    amount: NumberQuestions;
    difficulty?: GameDifficulty;
    type?: GameType;
}
