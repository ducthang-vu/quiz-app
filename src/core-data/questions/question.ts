import { GameDifficulty } from '../game-difficulty';
import { GameType } from '../game-type';

export interface Question {
    category: string;
    type: GameType;
    difficulty: GameDifficulty;
    question: string;
    correct_answer: string;
    incorrect_answers: string[];
}
