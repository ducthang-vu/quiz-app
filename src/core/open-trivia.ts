/**
 * Code 0: Success Returned results successfully.
 * Code 1: No Results Could not return results. The API doesn't have enough questions for your query. (Ex. Asking for 50 Questions in a Category that only has 20.)
 * Code 2: Invalid Parameter Contains an invalid parameter. Arguements passed in aren't valid. (Ex. Amount = Five)
 * Code 3: Token Not Found Session Token does not exist.
 * Code 4: Token Empty Session Token has returned all possible questions for the specified query. Resetting the Token is necessary.
 * Code 5: Rate Limit Too many requests have occurred. Each IP can only access the API once every 5 seconds.
 */
export enum ResponseCode {
    SUCCESS = 0,
    NO_RESULTS = 1,
    INVALID_PARAMETER = 2,
    TOKEN_NOT_FOUND = 3,
    TOKEN_EMPTY = 4,
    RATE_LIMIT = 5
}

export interface BaseResponse {
    response_code: ResponseCode;
}

export type GameDifficulty = 'any difficulty' | 'easy' | 'medium' | 'hard';

export type GameType = 'multiple' | 'boolean' | 'both';

export interface Question {
    category: string;
    type: GameType;
    difficulty: GameDifficulty;
    question: string;
    correct_answer: string;
    incorrect_answers: string[];
}

export interface QuestionsResponse extends BaseResponse {
    results: Question[];
}
