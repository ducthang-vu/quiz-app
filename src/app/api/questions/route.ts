import { NextRequest, NextResponse } from 'next/server';
import { BaseResponse, Question, QuestionsResponse, ResponseCode } from '@/core/open-trivia';
import { withRetry } from '@/lib/retry';

export interface TokenResponse extends BaseResponse {
    response_message: string;
    token: string;
}

const SERVER_ERROR_RESPONSE = NextResponse.json({ status: 500, message: 'Server Error' });

// TODO this only works in localhost, need a real db because vercel use lambda
let token: string | null = null;

async function getToken(): Promise<TokenResponse> {
    return fetch('https://opentdb.com/api_token.php?command=request').then(r => r.json() as unknown as TokenResponse);
}

function replaceQuotes(str: string): string {
    return str
        .replace(/&quot;/g, '"')
        .replace(/&amp;/g, '&')
        .replace(/&lt;/g, '<')
        .replace(/&gt;/g, '>')
        .replace(/&#39;/g, '"')
        .replace(/&#039;s/g, '')
        .replace(/&#039;/g, "'")
        .replace(/&eacute;/g, 'Ã©')
        .replace(/&rsquo;/g, "'")
        .replace(/&divide;/, '/');
}

async function _GET(req: NextRequest): Promise<NextResponse> {
    if (!token) {
        console.log('Fetching new token.');
        token = await getToken().then(r => r.token);
    }

    const amount = req.nextUrl.searchParams.get('amount');

    if (!amount || isNaN(Number(amount)) || Number(amount) <= 0) {
        return NextResponse.json({ status: 400, message: 'Invalid amount parameter' }, { status: 400 });
    }

    const response: Response = await fetch(`https://opentdb.com/api.php?amount=${amount}&token=${token}`);
    if (!response.ok) {
        return SERVER_ERROR_RESPONSE;
    }
    const body = await response.json() as QuestionsResponse;
    const tokenError = [ResponseCode.TOKEN_NOT_FOUND,  ResponseCode.TOKEN_EMPTY];
    if (tokenError.includes(body.response_code)) {
        token = null; // Reset token to trigger a new fetch on next request
        console.log('Token expired or not found, running again.');
        throw new Error('Token expired or not found');
    }
    const formattedBody = body.results.map((q: Question) => ({
        ...q,
        question: replaceQuotes(q.question),
        correct_answer: replaceQuotes(q.correct_answer),
        incorrect_answers: q.incorrect_answers.map(replaceQuotes),
    }));

    return NextResponse.json(formattedBody, { status: 200 });
}

export async function GET(req: NextRequest) {
    return withRetry(_GET)(req);
}
