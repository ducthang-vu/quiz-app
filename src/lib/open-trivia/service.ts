import { GetQuestionsParams, Question, QuestionsResponse, ResponseCode, TokenResponse } from '@/lib/open-trivia/types';
import { HttpError, OpenTriviaFailureError } from '@/lib/errors';

async function getToken(): Promise<string> {
    console.log('Fetching new token from OpenTrivia.');
    const response = await fetch('https://opentdb.com/api_token.php?command=request');
    const body = await response.json() as TokenResponse;
    if (!response.ok) {
        const error = new HttpError(`[OpenTriviaService#getToken] ${JSON.stringify(body)}`, response.status);
        console.error(error);
        throw error;
    }
    if (body.response_code !== ResponseCode.SUCCESS) {
        const error = new OpenTriviaFailureError(`[OpenTriviaService#getToken] Failed to get token: ${body.response_message}`, body.response_code);
        console.error(error);
        throw error;
    }

    return body.token;
}

async function getQuestions(payload: GetQuestionsParams): Promise<Question[]> {
    const res: Response =  await fetch(`https://opentdb.com/api.php?amount=${payload.amount}&token=${payload.token}`);
    const body = await res.json() as QuestionsResponse;
    if (!res.ok) {
        const error = new HttpError(JSON.stringify(body), res.status);
        console.error(error);
        throw error;
    }
    if (body.response_code !== ResponseCode.SUCCESS) {
        const error = new OpenTriviaFailureError(`[OpenTriviaService#getQuestions] Failed to get questions from OpenTrivia`, body.response_code);
        console.error(error);
        throw error;
    }
    return body.results;
}

export const openTriviaService = {
    getToken,
    getQuestions
}


