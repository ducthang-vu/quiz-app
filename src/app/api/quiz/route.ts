import { NextRequest, NextResponse } from 'next/server';
import { quizService } from '@/lib/quiz-service/service';
import { randomUUID } from 'node:crypto';
import { cookies } from 'next/headers';
import { COOKIE_NAME } from '@/app/api/constants';
import { GameDifficulty, GameType } from '@/lib/open-trivia/types';
import { CreateQuizPayload, QuizQuestion } from '@/lib/quiz-service/types';

function isGameType(value: string): value is GameType {
    return ['multiple', 'boolean', 'both'].includes(value);
}

function isGameDifficulty(value: string): value is GameDifficulty {
    return ['easy', 'medium', 'hard', 'any_difficulty'].includes(value);
}

export async function POST(req: NextRequest): Promise<NextResponse> {
    const body = await req.json();
    if (isNaN(Number(body.amount)) || Number(body.amount) <= 0) {
        return NextResponse.json({status: 400, message: 'Invalid amount parameter'}, {status: 400});
    }
    if (!isGameType(body.type)) {
        return NextResponse.json({ status: 400, message: 'Invalid type parameter' }, { status: 400 });
    }
    if (!isGameDifficulty(body.difficulty)) {
        return NextResponse.json({ status: 400, message: 'Invalid difficulty parameter' }, { status: 400 });
    }

    const cookie = await cookies();
    const id = cookie.get(COOKIE_NAME)?.value ?? randomUUID();
    cookie.set(COOKIE_NAME, id, { maxAge: 60 * 60 * 24 * 10 }); // Set cookie for 10 days

    const payload: CreateQuizPayload = {
        amount: Number(body.amount),
        difficulty: body.difficulty,
        type: body.type,
    }
    const questions: QuizQuestion[] = await quizService.createQuiz(id, payload);

    return NextResponse.json(questions, { status: 201 });
}
