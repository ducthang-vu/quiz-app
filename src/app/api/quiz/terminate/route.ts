import { cookies } from 'next/headers';
import { quizService } from '@/lib/quiz-service/service';
import { NextResponse } from 'next/server';
import { COOKIE_NAME } from '@/app/api/constants';

export async function POST(request: Request): Promise<NextResponse> {
    const id = await cookies().then(c => c.get(COOKIE_NAME)?.value);

    if (!id) {
        return NextResponse.json('Not found', { status: 404 });
    }

    const res: string[] = await request.json();
    await quizService.terminateQuiz(id, { answers: res })
    return NextResponse.json('ok', { status: 201 });
}
