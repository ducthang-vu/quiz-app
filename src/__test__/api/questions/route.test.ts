// @vitest-environment node

import { describe, expect, it, vi } from 'vitest';
import { NextRequest, NextResponse } from 'next/server';
import { GET, TokenResponse } from '@/app/api/questions/route';
import { ResponseCode } from '@/core/open-trivia';
import { NextURL } from 'next/dist/server/web/next-url';

const goodToken = 'good-token';

const tokenResponse: TokenResponse = {
    token: goodToken,
    response_message: '',
    response_code: ResponseCode.SUCCESS
}

// @ts-expect-error no need just for a test
vi.spyOn(global, "fetch").mockImplementation((url: string) => {
    if (url === 'https://opentdb.com/api_token.php?command=request') {
        // @ts-expect-error no need just for a test
        return Promise.resolve({ json() { return tokenResponse }})
    }
    if (url === 'https://opentdb.com/api.php?amount=2&token=good-token') {
        return Promise.resolve({
            ok: true,
            // @ts-expect-error no need just for a test
            json() {
                return {
                    response_code: ResponseCode.SUCCESS,
                    results: [
                        {
                            question: 'What is 2 + 2?',
                        },
                        {
                            question: 'What is the capital of France?',
                        }
                    ]
                };
            }
        });
    }
});

describe('route', () => {
    it('should return 400 when no param is given', async () => {
        const request = { nextUrl: {
                // eslint-disable-next-line @typescript-eslint/no-unused-vars
                searchParams: { get(name: string) { return null} } as URLSearchParams
            } as NextURL} as NextRequest;
        const response: NextResponse = await GET(request);
        expect(response.status).toEqual(400);
    });

    it('should return 200 with response', async () => {
        const request = {
            nextUrl: {
                searchParams: {
                    get(param: string) {
                        if (param === 'amount') {
                            return '2';
                        }
                        return null;
                    }
                } as URLSearchParams
            } as NextURL
        } as NextRequest;
        const response: NextResponse = await GET(request);
        expect(response.status).toEqual(200);
        expect(response.json()).toMatchSnapshot();
    });
})
