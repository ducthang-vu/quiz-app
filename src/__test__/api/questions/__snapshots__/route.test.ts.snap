// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`route > should return 200 with response 1`] = `Promise {}`;
